Add support for weighted-list faction templates.

martin509 opened the pull request:
## What type of PR is this?
1. [ ] Bug
2. [ ] Change
3. [x] Enhancement
4. [ ] Miscellaneous

### What have you changed and why?
Firstly, this adds support for constructing faction templates using weighted lists. See [https://community.bistudio.com/wiki/selectRandomWeighted](https://community.bistudio.com/wiki/selectRandomWeighted) for details. This allows template makers to create factions that spawn with an uneven distribution of weapons without making a large number of duplicate entries in the template, which is especially fun for more 'irregular' opponents such as SF and rivals. It also helps gameplay by giving factions common weapons that players can unlock easily, and rare weapons that can be a fun loot drop. This supports everything individual soldiers in a faction can spawn with, but not vehicles as implementing the same for vehicles would involve changes in a lot of different files.

In addition, this adds support for randomly choosing attachments (optics, suppressors, etc) in weapons by detecting if the weapon has a classname or an array, which simplifies faction templates by removing the number of permutations of a given weapon entry needed to have a variety of attachments - especially if e.g. multiple types of optic, accessory, etc all get mixed together. This functionality also supports weighted lists as well as regular arrays of classnames.

### Please specify which Issue this PR Resolves (If Applicable).
(N/A)

### Please verify the following.

1. [x] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?

1. [x] No
2. [ ] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps: Create or modify a faction to implements this feature (e.g. by adding weights to their weapon spawn lists, putting arrays in the attachment slots of their weapons, etc). An example of what this looks like in practice is below since it's easier than explaining in words.
![image](https://github.com/user-attachments/assets/ade78de6-4a9a-48e5-be14-c9b360bed216)
In this example, Mk20 rifles spawn at a 3:1 rate compared to TRG-21s, regular soldiers have ACO optics at a 1:4 rate compared to ironsights, team leaders have an equal distribution of ACOs, MRCOs and ironsights. When dealing with DLC compatibility, take care to make sure to keep track of item weights - if a spawn table is weighted, make sure to `append [item, weight]` instead of just `pushBack item`.
Once this is done, load the template in question in to a new Antistasi game and check the inventories of spawned soldiers.
********************************************************
Notes: I am currently working on modifying the vanilla AAF template to provide a thorough example of this implemented, but as that takes a long while I'm putting up this PR now.

martin509 committed:
Add support for weighted-list faction templates

wersal454 commented:
Needs testing, but this seems usefull

If someone had a time to combine it with mod_content and dlc_content files, would be even easier to create faction templates

martin509 commented:
> Needs testing, but this seems usefull
> 
> If someone had a time to combine it with mod_content and dlc_content files, would be even easier to create faction templates

The way things are implemented here, mod_content and dlc_content files are automatically supported since none of the handling code is template-specific - all it cares about is the data the template passes `fn_loadout_builder`.

Silence-112 reviewed (commented):
Looks good, i'll wait until the AAF conversion has been done before I approve

PetMudstone committed:
Start of weighted AAF faction files.

Includes integration files for DLCs.

wersal454 commented:
after this one, we might need to update example files

martin509 committed:
Add beginnings of AAF template.

martin509 committed:
Reweight most of the AAF config loot tables to add up to 10.

PetMudstone committed:
Partial weighting support for CSLA and GM.

Made before we decided to make default weights add up to 10 for all AAF equipment, so we will need to adjust weights again.

PetMudstone committed:
Merge remote-tracking branch 'upstream/unstable' into unstable

PetMudstone committed:
Quick fixes to Elite vanilla frequencies and gear.

Made sure heavy vest freqs summed up to 10, helmets actaully included an extra boonie instead of a redundant element, and also got rid of a redundant backpack list.

PetMudstone committed:
Making Militia equipment sum up to 10 (or close enough) properly.

Helmets sum up to 9 as to preserve ratio.

PetMudstone committed:
Rebalancing of GM and CSLA freqs, SF freq adjustment, Apex work.

Didn't touch DLC/CDLC freqs for SFs yet. Still need to make all SF vanilla freqs sum up to 10.

Removed DMRs from Apex compat file since they're all actually from Marksman.

PetMudstone committed:
Small freq tweaks, Marksman and Apex support.

martin509 committed:
Rationalizing AT launchers.

martin509 committed:
CUP AAF weapons, part 1.

PetMudstone committed:
Full Reaction Forces support, start of Western Sahara work.

Only covered SF weapons for WS thus far. Replaced almost all the weapons for SF with XMS variants since it doesn't make sense for 203X SF to be using FN FALs and Galils when a more obvious idiosyncratic option is present.

wersal454 commented:
@martin509 , @PetMudstone 

I really appreciate what you guys doing, weighted system will help a ton, but... 
in terms of gutting content out of templates themselves, dlc content or mod content(in this case CUP), I do not condole straight up deleting stuff. 
If you don't like something, you can always comment it out so when the time comes for a review, every addition/deletion of something could be easily restored/aborted or approved.
So before you've spent more time with this PR, I would suggest you reconsider.

martin509 commented:
> @martin509 , @PetMudstone
> 
> I really appreciate what you guys doing, weighted system will help a ton, but... in terms of gutting content out of templates themselves, dlc content or mod content(in this case CUP), I do not condole straight up deleting stuff. If you don't like something, you can always comment it out so when the time comes for a review, every addition/deletion of something could be easily restored/aborted or approved. So before you've spent more time with this PR, I would suggest you reconsider.

Totally fair to point this out - we were getting carried away because some of the content included in the template didn't quite make a lot of sense from a general faction alignment POV (for instance, why did basic AAF military get Desert Eagles with Reaction Forces). I'll undo the changes in question but probably will still tone down the frequencies on certain loot items that seem like they should be uncommon in AAF hands.

martin509 committed:
Revert back to old-style CUP weapons template

martin509 committed:
CUP all but elites done

PetMudstone committed:
Frequency support for SOG and Spearhead gear.

Doesn't include weapons.

PetMudstone committed:
Contact, Laws of War, and WS gear support. Fix RF gear file.

Also removed commando mortars from SF since i added them and that should be reviewed first.

PetMudstone committed:
SOG and Spearhead weapon weights.

wersal454 commented on A3A/addons/core/Templates/Templates/MOD_content/CUP/Vanilla_AAF/Weapons_AAF.sqf:
What you could do, is to cut 50 round mags from mk-17, mk-16 and mk-20. Leave them only in sl rifles. I think only squad leads should carry 50 mags.

martin509 replied:
I'm currently working through elites, so will do. Other changes to elites in the works:
- the most common guns are XM8, SCAR-L, and the 6.8mm ACR
- more unmagnified optics (eotech 553, M68 CCO) in the optics pool
- Green variants of guns outnumber black ones by ~2:1 since AAF
- bipods are not universal but instead have a ~1/3 or so chance to spawn
- currently considering adding at least the Mk20 and Mk16 SV to DMRs, since that particular area is lacking variety. Their commonality as regular rifles is also toned down a lot

wersal454 commented on A3A/addons/core/Templates/Templates/DLC_content/gear/Contact/Vanilla_AAF.sqf:
You could reduce the chances for CBRN suit. I couldn't reduce their spawn chance without weight system. I think this way, they make more sense

PetMudstone replied:
Will do, though we are going to audit all the weights in general after getting the general weight support done. We've been winging it initially due to the sheer quantity of items involved and have been working separately. A second pass should make the weights more cohesive.

PetMudstone committed:
Revert full deletions of items. Slight weight tweaking. Full WS support.

Still commented the M14s and SIGs in the Apex file because they are from the Marksman DLC and are already defined in the Marksman addon patch.

martin509 committed:
CUP elites complete

martin509 committed:
Minor item weight tweaks

martin509 committed:
Fix leaving out CUP optics lists

PetMudstone committed:
Give MK200/Stoner actual militia sights. Attachments for CUP militia.

PetMudstone committed:
Merge remote-tracking branch 'Martin/unstable' into unstable

martin509 committed:
Minor CUP config tweaks

martin509 committed:
CUP gear complete

martin509 committed:
CUP militia sights tweak

PetMudstone committed:
Adjust position of some CUP weapon lines to be nicer.

PetMudstone committed:
Merge remote-tracking branch 'Martin/unstable' into unstable

PetMudstone committed:
Adjust and add freqs for cosmetic goggles and glasses items.

PetMudstone committed:
Tons of formatting changes for readability and fixing some errors.

Added lengths of forward slashes to delineate sections for different tiers/groups and made gear lists with frequencies no longer be on a single line as they were getting even more crowded otherwise.
Also missed the Contact facewear previously.

wersal454 commented:
About SMGs, who should be equipped with them? Medics, AA/AT troops? Should they be equiped only SMGs or also with carbines or carbines and rifles?

martin509 commented:
> About SMGs, who should be equipped with them? Medics, AA/AT troops? Should they be equiped only SMGs or also with carbines or carbines and rifles?

As I understand it, in most templates they are pretty much exclusively for vehicle crews. Some templates (like SOG factions) have them mixed in either alongside or as a fallback if no carbines are defined, but in the AAF templates they're exclusively used by police and crewmen - medics and so on get carbines and every single subfaction has carbines defined.

wersal454 commented:
> > About SMGs, who should be equipped with them? Medics, AA/AT troops? Should they be equiped only SMGs or also with carbines or carbines and rifles?
> 
> As I understand it, in most templates they are pretty much exclusively for vehicle crews. Some templates (like SOG factions) have them mixed in either alongside or as a fallback if no carbines are defined, but in the AAF templates they're exclusively used by police and crewmen - medics and so on get carbines and every single subfaction has carbines defined.

yeahyeah, I meant if YOU would to add SMGs to some unit types, who would be equipped with them? Like, in all or almost all templates

martin509 commented:
> > > About SMGs, who should be equipped with them? Medics, AA/AT troops? Should they be equiped only SMGs or also with carbines or carbines and rifles?
> > 
> > 
> > As I understand it, in most templates they are pretty much exclusively for vehicle crews. Some templates (like SOG factions) have them mixed in either alongside or as a fallback if no carbines are defined, but in the AAF templates they're exclusively used by police and crewmen - medics and so on get carbines and every single subfaction has carbines defined.
> 
> yeahyeah, I meant if YOU would to add SMGs to some unit types, who would be equipped with them? Like, in all or almost all templates

I think the simplest implementation would be sprinkling SMGs into the carbine lists for militias since putting e.g., a random choice between the SMG list and the carbine list, in the actual soldier templates would force it to apply regardless of war level.

PetMudstone commented:
Hypothetically you could adjust the frequency subfactions or even unit types within subfactions get SMGs vs. carbines. Might lead to balancing headaches though. My pie in the sky idea would be that squads/units sent to defend or attack towns or cities would have SMGs used by riflemen but that's outside the scope of this PR and I doubt it would be an easy thing to implement at all.

martin509 committed:
Apex final review.

martin509 committed:
Art Of War final review.

martin509 committed:
GM final review.

martin509 committed:
SOG final review

martin509 committed:
Spearhead final review.

martin509 committed:
Contact final review.

martin509 committed:
CSLA final review.

martin509 committed:
Marksmen final review.

martin509 committed:
RF final review.

martin509 committed:
WS final review.

martin509 committed:
CUP final review.

martin509 committed:
minor optic rarity tweak

martin509 commented:
@SilenceIsFatto @wersal454 Pinging you to let you know that the AAF config (and every mod/DLC compat) is now complete. 

There are a ton of changes and I've tried to keep the general spirit and balance of the original template, but a number of weapons have been made extra rare where it seemed reasonable and the optics spawn lists are much more diverse to take full advantage of the feature, so there is unfortunately a fair amount to review. Nothing has been outright removed from spawn lists though, per Wersal's request.

wersal454 commented:
haven't tested it yet, but if you can, update FactionExample file

I think this is the PR to finally update it

martin509 commented:
> haven't tested it yet, but if you can, update FactionExample file
> 
> I think this is the PR to finally update it

Can you elaborate? Just add comments instructing how the weighted lists & etc work in that file?

wersal454 commented:
> > haven't tested it yet, but if you can, update FactionExample file
> > I think this is the PR to finally update it
> 
> Can you elaborate? Just add comments instructing how the weighted lists & etc work in that file?

Yeah, with an example.

@SilenceIsFatto maybe instead of an example we should replace old stuff with this in example file?

martin509 committed:
update FactionExample.sqf with instructions.

martin509 commented:
> > > haven't tested it yet, but if you can, update FactionExample file
> > > I think this is the PR to finally update it
> > 
> > 
> > Can you elaborate? Just add comments instructing how the weighted lists & etc work in that file?
> 
> Yeah, with an example.
> 
> @SilenceIsFatto maybe instead of an example we should replace old stuff with this in example file?

Done, there's a quick explanation in comments with examples of how the spawn lists and such work now.

wersal454 commented:
Nice, I'll hopefully try it during weekend

Also I will try to leave a review for example file with additions I wanted to add since summer, so there won't be a need for a separate PR just for this file

martin509 committed:
vanilla template bugfixes

martin509 committed:
bugfixes round 2

PetMudstone committed:
Bugfixes 3: Actually Working Builds Edition

PetMudstone committed:
Bugfixes 4: Non-critical misconfig fixes.

PetMudstone committed:
Bugfixes 5: The rest of the non-critical misconfig issues.

PetMudstone commented:
We've confirmed that the modified faction works, but are still waiting on the review @wersal454.

wersal454 commented:
optic_r1_high_black_lxWS doesn't exist

probably something I missed, needs to be fixed

wersal454 commented:
in CUP, weights with value more then 2 are bit much I think

I would reduce them to be 2 or lower, it just leads to a G36 and M16 bias

wersal454 commented:
other then that, good

![image](https://github.com/user-attachments/assets/72fafa75-c2f2-4793-9846-17cd83a9e090)

martin509 commented:
> in CUP, weights with value more then 2 are bit much I think
> 
> I would reduce them to be 2 or lower, it just leads to a G36 and M16 bias

I'll reduce the weights in rifle lists since those are fairly cluttered anyways, but I will note that in most cases the high weights are necessary to get some weapons to crowd others out since by default all the base-game item weights add up to 10. So for instance if a CUP weapon list adds up to 20 then overall there will be 2:1 CUP weapons to vanilla ones.

martin509 committed:
CUP weapon weight adjustment.

martin509 committed:
WS optic fix.

wersal454 commented:
ok, only this thing left and it's good to go
![image](https://github.com/user-attachments/assets/048c9a4c-f3d5-4d0b-9f9e-d2e992e0370c)

martin509 committed:
SOG vest tweak.

PetMudstone commented:
@wersal454 Everything should finally be properly ready.

Silence-112 commented on A3A/addons/core/Templates/Templates/DLC_content/gear/SOG/Vanilla_AAF.sqf:
```diff
@@ -1,38 +1,105 @@
+    "vn_g_spectacles_01", 1.5,
+    "vn_g_spectacles_02", 2
+    // "vn_b_spectacles",
+    // "vn_b_aviator"
```
Can remove these commented out classnames if they are redundant, will clean it up a bit

Silence-112 commented on A3A/addons/core/Templates/Templates/DLC_content/gear/WS/Vanilla_AAF.sqf:
```diff
@@ -1,18 +1,30 @@
+	"lxWS_H_PASGT_goggles_olive_F", 1.5,
+	"H_Beret_Headset_lxWS", 0.25
+];
+// (_eliteLoadoutData get "backpacks") pushBack "I_shield_backpack_lxWS"; // Isn't this a duplicate?
```
Seems like it, can be removed

Silence-112 commented on A3A/addons/core/Templates/Templates/DLC_content/weapons/Apex/Vanilla_AAF.sqf:
are `_sfAccessories`, `_eliteAccessories`, `_militaryAttachments` etc, ever defined anywhere in this file? Are they inherited from another file?

Seems to be this way in quite a few files, so if it is an issue then make sure to double check all of them

martin509 replied:
The variables are inherited from the base `Templates/Vanilla/Vanilla_AI_AAF.sqf`, in order to make e.g. adding DLC optics to vanilla guns much easier.

Silence-112 replied:
Makes sense, thanks for the clarification

Silence-112 commented on A3A/addons/core/Templates/Templates/DLC_content/weapons/Apex/Vanilla_AAF.sqf:
```diff
@@ -1,49 +1,52 @@
+//     ["srifle_DMR_03_khaki_F", "", "acc_pointer_IR", "optic_AMS_khk",["20Rnd_762x51_Mag","20Rnd_762x51_Mag","20Rnd_762x51_Mag"], [], "bipod_03_F_blk"],
+//     ["srifle_DMR_03_F", "", "acc_pointer_IR", "optic_AMS",["20Rnd_762x51_Mag","20Rnd_762x51_Mag","20Rnd_762x51_Mag"], [], "bipod_03_F_blk"],
+//     ["srifle_DMR_06_olive_F", "", "", "optic_AMS",["20Rnd_762x51_Mag","20Rnd_762x51_Mag","20Rnd_762x51_Mag"], [], "bipod_03_F_blk"]
+// ]; // Why are Marksman DLC items defined in the Apex file?
```
Good question, are they retextures added by apex?

If they have no relation and are solely from marksman, they can be removed

PetMudstone replied:
After checking, no they are not retextures and come wholly from the Marksman DLC.

Silence-112 reviewed (commented):
Generally speaking it looks fine, though I haven't had time to review the three biggest files. 

Left some feedback via review comments

PetMudstone committed:
Delete commented out entries after permission granted.

Remembered to also remove redundant shields.

PetMudstone committed:
Make Optic array names consistent.

Confirmed working after build made, no array related errors or warnings appeared.

PetMudstone committed:
Fuck CZ75 having itself as a magazine for some reason.

PetMudstone committed:
Fix CZ75 having itself as a magazine for some reason.

PetMudstone committed:
Merge remote-tracking branch 'Martin/unstable' into unstable

PetMudstone commented:
Should we also add the "PSRL-1" launcher (which comes in AAF camo), "Gendarmerie Tac Vest (Holster)", and "Mk26 (Stripes)" pistol (the RF version of the Deagle) that were added in the latest Reaction Forces update while we were working on this?

wersal454 commented:
> Should we also add the "PSRL-1" launcher (which comes in AAF camo), "Gendarmerie Tac Vest (Holster)", and "Mk26 (Stripes)" pistol (the RF version of the Deagle) that were added in the latest Reaction Forces update while we were working on this?

probably, though I think I added those in #426

PetMudstone commented:
Alright, should we wait until that gets merged before attempting further action, or will we wait for this to be merged considering we touch some of the same files?

wersal454 commented:
> Alright, should we wait until that gets merged before attempting further action, or will we wait for this to be merged considering we touch some of the same files?

By further action you mean, doing the same thing for other templates? I would suggest waiting for #426, since it has considerable amount of changes
(btw, maybe you guys could help by launching every template that was changed, I think I did that but maybe missed some)

Silence-112 reviewed (approved):
LGTM

Do you have any further work to do or can I add the ready for merge tag?

#426 touches far too many files to wait for

martin509 commented:
> LGTM
> 
> Do you have any further work to do or can I add the ready for merge tag?
> 
> #426 touches far too many files to wait for

Everything is ready to go!

wersal454 commented:
> > Should we also add the "PSRL-1" launcher (which comes in AAF camo), "Gendarmerie Tac Vest (Holster)", and "Mk26 (Stripes)" pistol (the RF version of the Deagle) that were added in the latest Reaction Forces update while we were working on this?
> 
> probably, though I think I added those in #426

After thinking about it, you should probably add new gear and weapons here after all, since Silence is probably going to push your PR first, I'll manage the conflict merges on my end

martin509 committed:
Reaction Forces 1.0.2 content

martin509 commented:
> > > Should we also add the "PSRL-1" launcher (which comes in AAF camo), "Gendarmerie Tac Vest (Holster)", and "Mk26 (Stripes)" pistol (the RF version of the Deagle) that were added in the latest Reaction Forces update while we were working on this?
> > 
> > 
> > probably, though I think I added those in #426
> 
> After thinking about it, you should probably add new gear and weapons here after all, since Silence is probably going to push your PR first, I'll manage the conflict merges on my end

Done! Everything is ready to merge.
